[JENKINS-58540] Mask secrets only if non-empty

karolgil opened the pull request:
[JENKINS-58540](https://issues.jenkins-ci.org/browse/JENKINS-58540)

Currently when secret is empty or is a single character SecretsMasker will not try to mask the value not to pollute logs with stream of asterisks.
User can override min length of secret that should be masked via 'org.csanchez.jenkins.plugins.kubernetes.pipeline.secretsMasking.minLength' property.

Property defaults to '2' to avoid masking empty secrets as well as single character secrets which may be guessed easily when all specific characters are
masked with '********' in logs.

karolgil commented:
@Vlatombe I must say I can't understand how it's over-engineering as comments in the issue already suggest some people would like to be able to specify the min length to mask.

I can of course change that. Just to be clear: you would like for only non-empty secrets to be masked, without the possibility to set minimum length, right?

Vlatombe commented:
@karolgil Yes. Anything more elaborate should be implemented in credentials-binding.

karolgil commented:
@Vlatombe fixed, let me know if that's sufficient.

Vlatombe commented:
@karolgil Thanks. Can you update the test code related to SecretMasker to verify that the output in case of empty secret is correct to prevent regressions?

jglick commented:
> this logic should be exposed by credential-binding plugin as an API usable by plugins

Yes, I need to file myself a reminder to do that.

jglick commented on src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/SecretsMasker.java:
```diff
@@ -71,19 +72,28 @@ public OutputStream decorate(OutputStream logger) throws IOException, Interrupte
             }
+
+            private String mask(String s, String value) {
+                if (value != null && value.length() > 0) {
```
```suggestion
                if (!value.isEmpty()) {
```

karolgil replied:
@jglick out of curiosity, are we sure we're not putting in any `null` values there? I couldn't find and promise like that, hence the null check.

jglick replied:
Manual analysis says no. It would be nice to `assert` that in the `SecretsMasker` constructor for clarity.

karolgil commented:
@jglick @Vlatombe I've added a `null` check for values to `SecretsMasker` constructor and added empty env vars to test - please let me know if that's correct, I'm not sure if I understood how that testing stuff work in here.

jglick commented on src/test/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/runWithEnvVariables.groovy:
```diff
@@ -25,6 +26,8 @@ podTemplate(
         echo OUTSIDE_POD_ENV_VAR = $POD_ENV_VAR
         echo OUTSIDE_POD_ENV_VAR_FROM_SECRET = $POD_ENV_VAR_FROM_SECRET or `echo $POD_ENV_VAR_FROM_SECRET | tr [a-z] [A-Z]`
+        echo OUTSIDE_EMPTY_POD_ENV_VAR = $EMPTY_POD_ENV_VAR
+        echo OUTSIDE_EMPTY_POD_ENV_VAR_FROM_SECRET = $EMPTY_POD_ENV_VAR_FROM_SECRET or `echo $EMPTY_POD_ENV_VAR_FROM_SECRET | tr [a-z] [A-Z]`
```
Do not really need to `tr` here, could just say something like

```sh
echo "OUTSIDE_EMPTY_POD_ENV_VAR_FROM_SECRET = '$EMPTY_POD_ENV_VAR_FROM_SECRET'"
```

which should print

```
OUTSIDE_EMPTY_POD_ENV_VAR_FROM_SECRET = ''
```

karolgil commented on src/test/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesTestUtil.java:
```diff
@@ -218,11 +219,17 @@ public static void createSecret(KubernetesClient client, String namespace) {
         secret = client.secrets().inNamespace(namespace).createOrReplace(secret);
         LOGGER.log(Level.INFO, "Created pod secret: {0}", secret);
+
+        secret = new SecretBuilder().withStringData(ImmutableMap.of(SECRET_KEY, EMPTY_POD_ENV_VAR_FROM_SECRET_VALUE))
```
Agreed, but I'm not sure how to do that properly without breaking making a mess. To verify this masking it needs to be called in one of `@Before` methods of  `AbstractKubernetesPipelineTest` and it already makes it a part of 5 test cases that derive from it. As cloud/secrets are set up in this class and not in test cases' classes I think it's more readable this way and having another secret here doesn't hurt imo.

jglick replied:
Yes, not a big deal if it would be a hassle to limit it to one test.

jglick commented:
By the way when you force-push changes to a PR, you break the GitHub feature which lets reviewers decide to look only at those changes since the last review. Generally I recommend simply pushing amending commits (no destructive history operations), and leave it to the repository maintainer to decide whether they want to squash, rebase, or just do a simple merge into the base branch.

jglick commented on src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/SecretsMasker.java:
```diff
@@ -54,6 +54,7 @@
     private static final Logger LOGGER = Logger.getLogger(SecretsMasker.class.getName());
 
     private static final long serialVersionUID = 1;
+    private static final String secretReplacement = "********";
```
No behavioral change, could be reverted.

jglick commented on src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/SecretsMasker.java:
```diff
@@ -71,19 +72,22 @@ public OutputStream decorate(OutputStream logger) throws IOException, Interrupte
                 }
                 logger.write(s.getBytes(StandardCharsets.UTF_8));
             }
+
```
Could be reverted.

jglick commented on src/test/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/runWithEnvVariables.groovy:
```diff
@@ -25,6 +26,7 @@ podTemplate(
         echo OUTSIDE_CONTAINER_ENV_VAR_FROM_SECRET = $CONTAINER_ENV_VAR_FROM_SECRET or `echo $CONTAINER_ENV_VAR_FROM_SECRET | tr [a-z] [A-Z]`
         echo OUTSIDE_POD_ENV_VAR = $POD_ENV_VAR
         echo OUTSIDE_POD_ENV_VAR_FROM_SECRET = $POD_ENV_VAR_FROM_SECRET or `echo $POD_ENV_VAR_FROM_SECRET | tr [a-z] [A-Z]`
+        echo OUTSIDE_EMPTY_POD_ENV_VAR_FROM_SECRET = '$EMPTY_POD_ENV_VAR_FROM_SECRET '
```
I think you meant

```suggestion
        echo "OUTSIDE_EMPTY_POD_ENV_VAR_FROM_SECRET = '$EMPTY_POD_ENV_VAR_FROM_SECRET'"
```

or

```suggestion
        echo OUTSIDE_EMPTY_POD_ENV_VAR_FROM_SECRET = "'$EMPTY_POD_ENV_VAR_FROM_SECRET'"
```

as what you have will print

```
OUTSIDE_EMPTY_POD_ENV_VAR_FROM_SECRET = '$EMPTY_POD_ENV_VAR_FROM_SECRET '
```

with no interpolation.

jglick commented:
Have a couple failures here in `KubernetesPipelineTest`.

Vlatombe commented on src/test/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/runWithEnvVariables.groovy:
```diff
@@ -2,6 +2,7 @@ podTemplate(
     envVars: [
         envVar(key: 'POD_ENV_VAR', value: 'pod-env-var-value'),
         secretEnvVar(key: 'POD_ENV_VAR_FROM_SECRET', secretName: 'pod-secret', secretKey: 'password')
```
```suggestion
        secretEnvVar(key: 'POD_ENV_VAR_FROM_SECRET', secretName: 'pod-secret', secretKey: 'password'),
```

Vlatombe commented:
Still one failure - https://kube.pse.beescloud.com/opensource/job/kubernetes-plugin/job/PR-558/4/testReport/org.csanchez.jenkins.plugins.kubernetes.pipeline/KubernetesPipelineTest/runInPodWithExistingTemplate/
